code/datums/components/scope.dm

```diff
@@ -0,0 +1,274 @@
+	var/tracker_owner_ckey
+	/// The method which we zoom in and out
+	var/zoom_method = ZOOM_METHOD_ITEM_ACTION
+	/// if not null, an item action will be added. Redundant if the mode is ZOOM_METHOD_RIGHT_CLICK or ZOOM_METHOD_WIELD.
+	var/item_action_type
+	/// Time to scope up, if you want the scope to take time to boot up. Used by the LWAP
+	var/time_to_scope
+	/// Are we currently trying to scope in?
+	var/attempting_to_scope = FALSE
+	/// Do we let the user scope and click on the middle of their screen?
+	var/allow_middle_click = FALSE
+
+/datum/component/scope/Initialize(range_modifier = 1, zoom_method = ZOOM_METHOD_ITEM_ACTION, item_action_type = /datum/action/zoom, time_to_scope = 0, allow_middle_click = FALSE)
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+	src.range_modifier = range_modifier
+	src.zoom_method = zoom_method
+	src.item_action_type = item_action_type
+	src.time_to_scope = time_to_scope
+	src.allow_middle_click = allow_middle_click
+
+
+/datum/component/scope/Destroy(force)
+	if(tracker)
+		stop_zooming(tracker.owner)
+	return ..()
+
+/datum/component/scope/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, PROC_REF(on_move))
+	switch(zoom_method)
+		if(ZOOM_METHOD_WIELD)
+			RegisterSignal(parent, SIGNAL_ADDTRAIT(TRAIT_WIELDED), PROC_REF(on_wielded))
+			RegisterSignal(parent, SIGNAL_REMOVETRAIT(TRAIT_WIELDED), PROC_REF(on_unwielded))
+	if(item_action_type)
+		var/obj/item/parent_item = parent
+		var/datum/action/scope = new item_action_type(parent)
+		parent_item.actions += scope
+		RegisterSignal(scope, COMSIG_ACTION_TRIGGER, PROC_REF(on_action_trigger))
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, PROC_REF(on_examine))
+	if(istype(parent, /obj/item/gun))
+		RegisterSignal(parent, COMSIG_GUN_TRY_FIRE, PROC_REF(on_gun_fire))
+
+/datum/component/scope/UnregisterFromParent()
+	if(item_action_type)
+		var/obj/item/parent_item = parent
+		var/datum/action/scope = locate(item_action_type) in parent_item.actions
+		parent_item.actions -= (scope)
+	UnregisterSignal(parent, list(
+		COMSIG_MOVABLE_MOVED,
+		SIGNAL_ADDTRAIT(TRAIT_WIELDED),
+		SIGNAL_REMOVETRAIT(TRAIT_WIELDED),
+		COMSIG_GUN_TRY_FIRE,
+		COMSIG_PARENT_EXAMINE,
+	))
+
+/datum/component/scope/process()
+	var/mob/user_mob = tracker.owner
+	var/client/user_client = user_mob.client
+	if(!user_client)
+		stop_zooming(user_mob)
+		return
+	tracker.calculate_params()
+	animate(user_client, world.tick_lag, pixel_x = tracker.given_x, pixel_y = tracker.given_y)
+
+/datum/component/scope/proc/on_move(atom/movable/source, atom/oldloc, dir, forced)
+	SIGNAL_HANDLER
+
+	if(!tracker)
+		return
+	stop_zooming(tracker.owner)
+
+/datum/component/scope/proc/on_action_trigger(datum/action/source)
+	SIGNAL_HANDLER
+	var/obj/item/item = source.target
+	var/mob/living/user = item.loc
+	if(is_internal_organ(item))
+		var/obj/item/organ/internal/O = item
+		user = O.owner
+	if(tracker)
```

Comment:
Might be good to add a proc `is_zoomed_in` which returns `!!tracker` to make this check a bit more semantical. Now it is not 100% clear why this check is here.

Reply:
done